File: general/src/main/java/com/github/ummo93/game/FighterShip.java
package com.github.ummo93.game;

import com.github.ummo93.framework.ActorAnimated2D;
import com.github.ummo93.framework.ActorTexture2D;
import com.github.ummo93.framework.AnimatedTexture;
import com.github.ummo93.framework.service.TaskQueueService;
import lombok.Getter;
import lombok.Setter;

import java.util.ArrayDeque;
import java.util.HashSet;
import java.util.Queue;
import java.util.Set;

import static com.github.ummo93.game.ControlSignal2D.*;
import static com.github.ummo93.utils.RaylibUtils.*;
import static com.github.ummo93.utils.RaylibUtils.vector2;
import static com.raylib.Raylib.*;
import com.raylib.*;


public class FighterShip extends ActorTexture2D implements Controllable, Damagable {

    private final Set<ControlSignal2D> controlSignals = new HashSet<>();
    private AnimatedTexture explosionAnimation;

    @Getter
    private final Queue<Bullet> bullets = new ArrayDeque<>();

    @Getter
    @Setter
    private Vector2 velocity = vector2Zero();
    @Getter
    @Setter
    private float rotationVelocity = 0;
    @Getter
    private final float speed = .1f;
    @Getter
    private final float maxSpeed = 50.f;
    @Getter
    private final float damage = 25.f;
    @Getter
    private double hp = 100.0;
    @Getter
    private final double maxHp = 100.0;
    @Getter
    private final int maxBullets = 10;
    @Getter
    private final float bulletSpeed = 15.f;
    @Getter
    private final float distanceToBulletDestroy = 800.f;
    @Getter
    boolean isForwardEngineActive = false;
    @Getter
    boolean isBackwardEngineActive = false;
    private Sound killSound;
    private Sound shootSound;
    private Texture explosionTexture;

    public void addCollider() {
        var collider = new BoundingBox()
            .max(vector3(position.x() + texture.width()/2.f, position.y() + texture.height()/2.f, 0.f))
            .min(vector3(position.x() - texture.width()/2.f, position.y() - texture.height()/2.f, 0.f));
        setCollider(collider);
    }

    public FighterShip(Vector3 position, Vector3 rotation, Texture texture) {
        super(position, rotation, texture);
        addCollider();
    }
    public FighterShip(Texture texture) {
        super(texture);
        addCollider();
    }

    @Override
    public void moveForward() {
        controlSignals.add(MOVE_FORWARD);
    }

    @Override
    public void moveBackward() {
        controlSignals.add(MOVE_BACKWARD);
    }

    @Override
    public void rotateClockwise() {
        controlSignals.add(ROTATE_CLOCKWISE);
    }

    @Override
    public void rotateCounterClockwise() {
        controlSignals.add(ROTATE_COUNTERCLOCKWISE);
    }

    @Override
    public void shoot() {
        controlSignals.add(SHOOT);
        playSound(shootSound);
    }

    @Override
    public void addDamage(double damage) {
        hp = Math.clamp(hp - damage, 0.0, maxHp);
        if (hp == 0) {
            getScene().remove(this);
        }
    }

    @Override
    protected void onInit() {
        explosionTexture = loadTextureResource("explosion-1.png");
        explosionAnimation = new AnimatedTexture(explosionTexture, 32, 32, 0, 7, 7, 1);
        killSound = loadSoundResource("blast.wav");
        shootSound = loadSoundResource("lazer.wav");
        super.onInit();
    }

    @Override
    protected void onUpdate(float dt) {
        var newRotation = calculateRotation();
        var currentRotation = getRotation();
        setRotation(vector3(currentRotation.x(), currentRotation.y() + (dt*45f*newRotation), currentRotation.z()));
        setVelocity(calculateVelocity());
        var newTranslatedPos = translate2D(vector2(position), vector2Scale(velocity, dt*75f*speed), 1f);
        setPosition(vector3(newTranslatedPos));

        if (controlSignals.contains(SHOOT)) {
            emitBullet();
        }

        refreshBulletsState(dt);
        controlSignals.clear();
    }

    @Override
    protected void onDraw() {
        if (isForwardEngineActive) {
            drawForwardExhaust(vector2(position), getForward2D());
        }
        if (isBackwardEngineActive) {
            drawBackwardExhaust(vector2(position), getForward2D());
        }

        drawBullets();
        drawHeadingVector();
        super.onDraw();
    }

    @Override
    protected void onDestroy() {
        var explosionPoint = vector3SubtractValue(position, explosionAnimation.getFrameWidth()/2f);
        var explosion = new ActorAnimated2D(explosionPoint, rotation, explosionAnimation);
        var scene = getScene();
        scene.spawn(explosion);
        TaskQueueService.getInstance().enqueue(() -> {
            scene.remove(explosion);
            unloadTexture(explosionTexture);
            unloadSound(killSound);
        }, 1);
        unloadSound(shootSound);
        playSound(killSound);
        super.onDestroy();
    }

    protected void emitBullet() {
        if (bullets.size() > maxBullets - 1) {
            bullets.remove();
        }
        bullets.add(new Bullet(vector2(position), getForward2D()));
    }

    protected void drawBullets() {
        for (var bullet : bullets) {
            drawLineEx(bullet.pos, vector2Add(bullet.pos, vector2Scale(bullet.fwd, 8)), 2, RED);
        }
    }

    protected void refreshBulletsState(float dt) {
        if (bullets.isEmpty()) return;
        for (var bullet : bullets) {
            bullet.pos = translate2D(vector2(bullet.pos.x(), bullet.pos.y()), vector2Scale(bullet.fwd, dt*800f), 1.f);
            if (vector2Distance(bullet.pos, vector2(position)) > distanceToBulletDestroy) {
                bullets.remove(bullet);
                break;
            }
            Ray ray = ray(bullet.pos, bullet.fwd);
            var infoOpt = getScene().raycastOne(ray, 10, this);
            if (infoOpt.isPresent()) {
                var info = infoOpt.get();
                var other = info.getOther();
                bullets.remove(bullet);
                if (other instanceof Damagable damagable) {
                    damagable.addDamage(damage);
                }
                break;
            }
        }

    }

    private void drawHeadingVector() {
        var endPos = translate2D(vector2(position), velocity, 1);
        drawText("<>", (int)(endPos.x()) - 5, (int)(endPos.y() ) - 5, 14, DARKGREEN);
    }

    private void drawForwardExhaust(Vector2 origin, Vector2 fwd) {
        var backward = vector2Negate(fwd);
        var endPosBezier = translate2D(origin, backward, getRandomValue(15, 25));
        var endPosBezier2 = translate2D(origin, backward, getRandomValue(20, 30));
        drawLineEx(endPosBezier, origin, 6.f, color(112, 31, 126, 150));
        drawLineEx(endPosBezier2, origin, 2.f, color(0, 121, 241, 200));
    }

    private void drawBackwardExhaust(Vector2 origin, Vector2 fwd) {
        var endPosBezier = translate2D(origin, fwd, getRandomValue(8, 13));
        var endPosBezier2 = translate2D(origin, fwd, 13);
        drawLineEx(endPosBezier, origin, 18.f, color(0, 121, 241, 150));
        drawLineEx(endPosBezier2, origin, 14.f, BLACK);
    }

    private float calculateRotation() {
        final float rotationSpeed = 0.5f;
        final float rotationMaxSpeed = 4.f;
        if (controlSignals.contains(ROTATE_COUNTERCLOCKWISE)) {
            rotationVelocity = clamp(rotationVelocity - rotationSpeed, -rotationMaxSpeed, rotationMaxSpeed);
        } else if (controlSignals.contains(ROTATE_CLOCKWISE)) {
            rotationVelocity = clamp(rotationVelocity + rotationSpeed, -rotationMaxSpeed, rotationMaxSpeed);
        } else {
            if (Math.abs(rotationVelocity) > 0.05f) {
                if (rotationVelocity > 0) {
                    rotationVelocity = clamp(rotationVelocity - 0.1f, -rotationMaxSpeed, rotationMaxSpeed);
                } else {
                    rotationVelocity = clamp(rotationVelocity + 0.1f, -rotationMaxSpeed, rotationMaxSpeed);
                }
            }
        }
        return rotationVelocity;
    }

    private Vector2 calculateVelocity() {
        var fwd = getForward2D();
        var newVelocity = velocity;
        if (controlSignals.contains(MOVE_FORWARD)) {
            isForwardEngineActive = true;
            isBackwardEngineActive = false;
            newVelocity = vector2Add(newVelocity, fwd);
            if (vector2Length(newVelocity) > maxSpeed) {
                newVelocity = vector2Scale(vector2Normalize(newVelocity), maxSpeed);
            }
        } else if (controlSignals.contains(MOVE_BACKWARD)) {
            isForwardEngineActive = false;
            isBackwardEngineActive = true;
            newVelocity = vector2Add(newVelocity, vector2Negate(fwd));
            if (vector2Length(newVelocity) > maxSpeed) {
                newVelocity = vector2Scale(vector2Normalize(newVelocity), maxSpeed);
            }
        } else {
            isForwardEngineActive = false;
            isBackwardEngineActive = false;
            if (vector2Length(newVelocity) > 0.25f) {
                newVelocity = vector2Scale(newVelocity, 0.996f);
            } else {
                newVelocity = vector2(0, 0);
            }
        }
        return newVelocity;
    }
}

File: sample-game/src/main/kotlin/com/github/ummo93/game/FighterShip.kt
package com.github.ummo93.game

import com.github.ummo93.framework.AnimatedTexture
import com.github.ummo93.framework.actors.AnimatedActor
import com.github.ummo93.framework.actors.TexturedActor
import com.github.ummo93.framework.service.TaskQueueService
import com.github.ummo93.framework.utils.RaylibUtils
import com.raylib.Colors
import com.raylib.Jaylib
import com.raylib.Raylib.*
import java.util.*
import kotlin.math.abs

class FighterShip : TexturedActor, Controllable, Damagable {
    private val controlSignals: MutableSet<ControlSignal2D> = HashSet()
    private var explosionAnimation: AnimatedTexture? = null

    val bullets: Queue<Bullet> = ArrayDeque()

    var velocity: Vector2 = Jaylib.vector2Zero()

    var rotationVelocity = 0f

    val speed = .1f

    val maxSpeed = 50f

    val damage = 25f

    override var hp: Double = 100.0

    val maxHp: Float = 100.0f

    val maxBullets = 10

    val bulletSpeed = 15f

    val distanceToBulletDestroy = 800f

    var isForwardEngineActive: Boolean = false

    var isBackwardEngineActive: Boolean = false

    private var killSound: Sound? = null
    private var shootSound: Sound? = null
    private var explosionTexture: Texture? = null

    fun addCollider() {
        val collider = BoundingBox()
            .max(RaylibUtils.vector3(position.x() + texture.width() / 2f, position.y() + texture.height() / 2f, 0f))
            .min(RaylibUtils.vector3(position.x() - texture.width() / 2f, position.y() - texture.height() / 2f, 0f))
        setCollider(collider)
    }

    constructor(position: Vector2?, rotation: Vector2?, texture: Texture?) : super(position, rotation, texture) {
        addCollider()
    }

    constructor(texture: Texture?) : super(texture) {
        addCollider()
    }

    override fun moveForward() {
        controlSignals.add(ControlSignal2D.MOVE_FORWARD)
    }

    override fun moveBackward() {
        controlSignals.add(ControlSignal2D.MOVE_BACKWARD)
    }

    override fun rotateClockwise() {
        controlSignals.add(ControlSignal2D.ROTATE_CLOCKWISE)
    }

    override fun rotateCounterClockwise() {
        controlSignals.add(ControlSignal2D.ROTATE_COUNTERCLOCKWISE)
    }

    override fun shoot() {
        controlSignals.add(ControlSignal2D.SHOOT)
        Jaylib.playSound(shootSound)
    }

    override fun addDamage(damage: Double) {
        hp = Math.clamp(hp - damage, 0.0, maxHp.toDouble())
        if (hp == 0.0) {
            getScene().remove(this)
        }
    }

    override fun onInit() {
        explosionTexture = RaylibUtils.loadTextureResource("explosion-1.png")
        explosionAnimation = AnimatedTexture(explosionTexture!!, 32, 32, 0, 7, 7, 1f)
        killSound = RaylibUtils.loadSoundResource("blast.wav")
        shootSound = RaylibUtils.loadSoundResource("lazer.wav")
        super.onInit()
    }

    override fun onUpdate(dt: Float) {
        val newRotation = calculateRotation()
        val currentRotation = getRotation()
        setRotation(RaylibUtils.vector2(currentRotation.x(), currentRotation.y() + (dt * 45f * newRotation)))
        velocity = calculateVelocity()
        val newTranslatedPos = RaylibUtils.translate2D(position, Jaylib.vector2Scale(velocity, dt * 75f * speed), 1f)
        setPosition(newTranslatedPos)

        if (controlSignals.contains(ControlSignal2D.SHOOT)) {
            emitBullet()
        }

        refreshBulletsState(dt)
        controlSignals.clear()
    }

    override fun onDraw() {
        if (isForwardEngineActive) {
            drawForwardExhaust(position, forward)
        }
        if (isBackwardEngineActive) {
            drawBackwardExhaust(position, forward)
        }

        drawBullets()
        drawHeadingVector()
        super.onDraw()
    }

    override fun onDestroy() {
        val explosionPoint = Jaylib.vector2SubtractValue(position, explosionAnimation!!.frameWidth / 2f)
        val explosion = AnimatedActor(explosionPoint, rotation, explosionAnimation)
        val scene = getScene()
        scene.spawn(explosion)
        TaskQueueService.getInstance().enqueue({
            scene.remove(explosion)
            Jaylib.unloadTexture(explosionTexture)
            Jaylib.unloadSound(killSound)
        }, 1.0)
        Jaylib.unloadSound(shootSound)
        Jaylib.playSound(killSound)
        super.onDestroy()
    }

    fun emitBullet() {
        if (bullets.size > maxBullets - 1) {
            bullets.remove()
        }
        bullets.add(Bullet(position, forward))
    }

    protected fun drawBullets() {
        for (bullet in bullets) {
            Jaylib.drawLineEx(
                bullet.pos,
                Jaylib.vector2Add(bullet.pos, Jaylib.vector2Scale(bullet.fwd, 8f)),
                2f,
                Colors.RED
            )
        }
    }

    protected fun refreshBulletsState(dt: Float) {
        if (bullets.isEmpty()) return
        for (bullet in bullets) {
            bullet.pos = RaylibUtils.translate2D(bullet.pos, Jaylib.vector2Scale(bullet.fwd, dt * 800f), 1f)
            if (Jaylib.vector2Distance(bullet.pos, position) > distanceToBulletDestroy) {
                bullets.remove(bullet)
                break
            }
            val ray = RaylibUtils.ray(bullet.pos, bullet.fwd)
            val infoOpt = getScene().raycastOne(ray, 10f, this)
            if (infoOpt.isPresent) {
                val info = infoOpt.get()
                val other = info.other
                bullets.remove(bullet)
                if (other is Damagable) {
                    other.addDamage(damage.toDouble())
                }
                break
            }
        }
    }

    private fun drawHeadingVector() {
        val endPos = RaylibUtils.translate2D(position, velocity, 1f)
        Jaylib.drawText("<>", (endPos.x()).toInt() - 5, (endPos.y()).toInt() - 5, 14, Colors.DARKGREEN)
    }

    private fun drawForwardExhaust(origin: Vector2, fwd: Vector2) {
        val backward = Jaylib.vector2Negate(fwd)
        val endPosBezier = RaylibUtils.translate2D(origin, backward, Jaylib.getRandomValue(15, 25).toFloat())
        val endPosBezier2 = RaylibUtils.translate2D(origin, backward, Jaylib.getRandomValue(20, 30).toFloat())
        Jaylib.drawLineEx(endPosBezier, origin, 6f, RaylibUtils.color(112, 31, 126, 150))
        Jaylib.drawLineEx(endPosBezier2, origin, 2f, RaylibUtils.color(0, 121, 241, 200))
    }

    private fun drawBackwardExhaust(origin: Vector2, fwd: Vector2) {
        val endPosBezier = RaylibUtils.translate2D(origin, fwd, Jaylib.getRandomValue(8, 13).toFloat())
        val endPosBezier2 = RaylibUtils.translate2D(origin, fwd, 13f)
        Jaylib.drawLineEx(endPosBezier, origin, 18f, RaylibUtils.color(0, 121, 241, 150))
        Jaylib.drawLineEx(endPosBezier2, origin, 14f, Colors.BLACK)
    }

    private fun calculateRotation(): Float {
        val rotationSpeed = 0.5f
        val rotationMaxSpeed = 4f
        if (controlSignals.contains(ControlSignal2D.ROTATE_COUNTERCLOCKWISE)) {
            rotationVelocity = Jaylib.clamp(rotationVelocity - rotationSpeed, -rotationMaxSpeed, rotationMaxSpeed)
        } else if (controlSignals.contains(ControlSignal2D.ROTATE_CLOCKWISE)) {
            rotationVelocity = Jaylib.clamp(rotationVelocity + rotationSpeed, -rotationMaxSpeed, rotationMaxSpeed)
        } else {
            if (abs(rotationVelocity.toDouble()) > 0.05f) {
                rotationVelocity = if (rotationVelocity > 0) {
                    Jaylib.clamp(rotationVelocity - 0.1f, -rotationMaxSpeed, rotationMaxSpeed)
                } else {
                    Jaylib.clamp(rotationVelocity + 0.1f, -rotationMaxSpeed, rotationMaxSpeed)
                }
            }
        }
        return rotationVelocity
    }

    private fun calculateVelocity(): Vector2 {
        val fwd = forward
        var newVelocity = velocity
        if (controlSignals.contains(ControlSignal2D.MOVE_FORWARD)) {
            isForwardEngineActive = true
            isBackwardEngineActive = false
            newVelocity = Jaylib.vector2Add(newVelocity, fwd)
            if (Jaylib.vector2Length(newVelocity) > maxSpeed) {
                newVelocity = Jaylib.vector2Scale(Jaylib.vector2Normalize(newVelocity), maxSpeed)
            }
        } else if (controlSignals.contains(ControlSignal2D.MOVE_BACKWARD)) {
            isForwardEngineActive = false
            isBackwardEngineActive = true
            newVelocity = Jaylib.vector2Add(newVelocity, Jaylib.vector2Negate(fwd))
            if (Jaylib.vector2Length(newVelocity) > maxSpeed) {
                newVelocity = Jaylib.vector2Scale(Jaylib.vector2Normalize(newVelocity), maxSpeed)
            }
        } else {
            isForwardEngineActive = false
            isBackwardEngineActive = false
            newVelocity = if (Jaylib.vector2Length(newVelocity) > 0.25f) {
                Jaylib.vector2Scale(newVelocity, 0.996f)
            } else {
                RaylibUtils.vector2(0f, 0f)
            }
        }
        return newVelocity
    }
}
